### PR TITLE
fix(cli): gracefully handle 403 errors when publishing

### PR DESCRIPTION
**Pull request by emmenko**
Ref #478

This is a (naive) attempt to gracefully handle 403 errors.

Let me know if this is an acceptable approach or if we should go into another direction.

### CODE REVIEW

**Comment by emmenko** [comment]
Has anyone had a chance to look at this? Let me know in case you want to do it differently.
cc @Andarist 

It would be nice to fix this as it's still causing some issues.

File: packages/cli/src/commands/publish/npm-utils.ts [hunk 1]
```diff
@@ -183,6 +185,31 @@ async function internalPublish(
...
+      json.error.code === "E403" &&
+      json.error.summary.includes("previously published versions")
+    ) {
+      let packageJson = await fs.readJSON(path.join(opts.cwd, "package.json"), {
+        encoding: "utf8"
+      });
+      let publishedPkgInfo = await getPackageInfo(
+        `${pkgName}@${packageJson.version}`
+      );
+      if (publishedPkgInfo.version === packageJson.version) {
```
Comment (Andarist) [inline packages/cli/src/commands/publish/npm-utils.ts:208]: what's the role of this check? shouldnt we trust the returned E403 that this particular version is already published?

**Comment by Andarist** [comment]
From what I understand this can occur during a run that happens after a failing run that actually has successfully published a package? Is that statement correct?

**Comment by emmenko** [comment]
Thanks for getting back to me.

The recap the issue: when publishing the packages, sometimes NPM returns `403 Forbidden` errors.

For example (see also CI run https://github.com/commercetools/ui-kit/runs/1736041131):

```
🦋  error an error occurred while publishing @commercetools-uikit/date-field: E403 403 Forbidden - PUT https://registry.npmjs.org/@commercetools-uikit%2fdate-field - You cannot publish over the previously published versions: 10.44.0. 
🦋  error In most cases, you or one of your dependencies are requesting
🦋  error a package version that is forbidden by your security policy.
🦋  error an error occurred while publishing @commercetools-uikit/async-select-input: E403 403 Forbidden - PUT https://registry.npmjs.org/@commercetools-uikit%2fasync-select-input - You cannot publish over the previously published versions: 10.44.0. 
🦋  error In most cases, you or one of your dependencies are requesting
🦋  error a package version that is forbidden by your security policy.
🦋  error an error occurred while publishing @commercetools-uikit/date-time-input: E403 403 Forbidden - PUT https://registry.npmjs.org/@commercetools-uikit%2fdate-time-input - You cannot publish over the previously published versions: 10.44.0. 
🦋  error In most cases, you or one of your dependencies are requesting
🦋  error a package version that is forbidden by your security policy.
```

This leads to changesets to not complete the publishing action, AND to not create git tags.

```
🦋  error packages failed to publish:
🦋  @commercetools-uikit/date-field@10.44.0
🦋  @commercetools-uikit/async-select-input@10.44.0
🦋  @commercetools-uikit/date-time-input@10.44.0
```

However, the packages are indeed published to NPM. If you check those packages listed above, the version `10.44.0` is published.

image

Therefore, the proposal in this PR is to "catch" those 403 errors and make sure that the version is indeed published. If that's the case, we can "safely" ignore the error.

PS: also worth mentioning that I have a feeling that these issues occur more frequently when lots of packages need to be published.

**Comment by NMinhNguyen** [comment]
That's very interesting given that there is logic in place to check whether a package needs to be published. Perhaps there is an edge case that the check doesn't account for? https://github.com/atlassian/changesets/blob/049f0b3122f506faec857a9c36c5f4b5cc550d67/packages/cli/src/commands/publish/publishPackages.ts#L191-L193

**Comment by Andarist** [comment]
Well, according to @emmenko and logs in that linked run this was the very first time that those versions were being published. So the initial check wasn't able to "discard" those from being published as they were not published at that time. 

It's super odd that a successful publish results in a 403 with such an error message. I've asked around some other people that are managing large monorepos to check if they have experienced this.

I don't see anything in particular wrong with the proposed approach - other than the fact that this shouldn't be needed, at least in theory. I would like to propose something different that I've wanted to do anyway and that might help to mitigate the issue - we should change the concurrency limit for publishing to 1.

If that won't help and this will still happen then I will feel more comfortable with using the logic proposed in this PR. WDYT?

**Comment by emmenko** [comment]
> we should change the concurrency limit for publishing to 1.

Yes that would be another option. 

Would this be the place for limiting the concurrent requests?

https://github.com/atlassian/changesets/blob/12f9a433a6c3ac38f9405fcd77c9108c423d7101/packages/cli/src/commands/publish/publishPackages.ts#L92

Or did you have something else in mind? Also, should the concurrency option be configurable?

**Comment by emmenko** [comment]
@Andarist I went ahead of opened a separate PR with the concurrency limit, as you suggested.
https://github.com/atlassian/changesets/pull/522

Let me know if this is what you had in mind. Thanks

**Comment by Andarist** [comment]
For now let's close this one as the other PR has landed: https://github.com/atlassian/changesets/pull/522 . We can always reopen this in the future if needed.

**Pull request closed without merging**